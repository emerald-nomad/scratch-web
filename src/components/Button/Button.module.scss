@import "styles/mixins";

.btn-container {
  display: flex;
  flex-direction: column;
  margin: 3em 0;

  @include mq(md) {
    flex-direction: row;
    justify-content: center;
  }
}

.btn {
  font-size: 1.25rem;
  margin: 0 1.5em;
  border: none;
  border-radius: 5px;
  padding: 0.75em 1.5em;
  flex: 1;

  @include mq(md) {
    padding: 0.5em 1.5em;
    margin: 0;
  }

  & + & {
    margin-top: 1.5em;

    @include mq(md) {
      margin-top: 0;
      margin-left: 1.5em;
    }
  }

  &--solid {
    background-color: var(--green);
    color: var(--white);
  }

  &--outline {
    background-color: var(--white);
    color: var(--green);
    border: 1px solid var(--green);
  }
}
