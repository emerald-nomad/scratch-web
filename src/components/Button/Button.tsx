import Link from "next/link";
import { ButtonHTMLAttributes } from "react";
import styles from "./Button.module.scss";

export interface ButtonProps {
  text: string;
  style?: "solid" | "outline" | "text";
  href?: string;
  onClick?: () => void;
  type?: ButtonHTMLAttributes<HTMLButtonElement>["type"];
}

const Button: React.FC<ButtonProps> = ({
  text,
  style = "solid",
  href,
  type = "button",
  onClick,
}) => {
  const Btn = (
    <button
      onClick={onClick}
      className={[styles.btn, styles[`btn--${style}`]].join(" ")}
      type={type}
    >
      {text}
    </button>
  );

  return href ? <Link href={href}>{Btn}</Link> : Btn;
};

export default Button;
