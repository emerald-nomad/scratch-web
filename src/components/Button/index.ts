export {default as Button} from "./Button"
export {default as ButtonContainer} from "./ButtonContainer"