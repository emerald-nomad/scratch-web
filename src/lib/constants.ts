export const APP_SECRET = process.env.APP_SECRET as string;
export const MONGO_URL = process.env.MONGO_URL as string;
export const API_URL = process.env.API_URL as string;