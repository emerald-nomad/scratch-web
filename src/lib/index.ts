export * from "./apolloClient"
export * from "./constants"