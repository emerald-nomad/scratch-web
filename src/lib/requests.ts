import { gql } from "@apollo/client";

export const LOGIN_USER = gql`
  mutation Login($input: LoginInput!) {
    login(input: $input) {
      token
      user {
        id
        name
        username
      }
    }
  }
`;

export const SIGN_UP_USER = gql`
  mutation SignUp($input: SignUpInput!) {
    signup(input: $input) {
      token
      user {
        id
        name
        username
      }
    }
  }
`;
