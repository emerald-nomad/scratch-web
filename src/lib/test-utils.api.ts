import { ApolloServer, ServerInfo } from "apollo-server";
import getPort, { makeRange } from "get-port";
import { GraphQLClient } from "graphql-request";
import { fieldAuthorizePlugin, makeSchema } from "nexus";
import * as types from "graphql/types";
import { AuthorizationError } from "graphql/errors";

import { join } from "path";
import { mockAuthRepository, mockUserRepository } from "graphql/repositories";

const testServer = new ApolloServer({
  schema: makeSchema({
    types,
    plugins: [
      fieldAuthorizePlugin({ formatError: () => AuthorizationError() }),
    ],
    contextType: {
      module: join(process.cwd(), "src", "graphql", "./context.ts"),
      export: "Context",
    },
    outputs: {
      typegen: join(process.cwd(), "generated/nexus-typegen.ts"),
      schema: join(process.cwd(), "generated/schema.graphql"),
    },
  }),
  context: (ctx: any) => ({
    ...ctx,
    Auth: mockAuthRepository,
    User: mockUserRepository,
  }),
});

type TestContext = {
  client: GraphQLClient;
};

export function createTestContext(): TestContext {
  let ctx = {} as TestContext;
  const graphqlCtx = graphqlTestContext();

  beforeEach(async () => {
    const client = await graphqlCtx.before();
    Object.assign(ctx, {
      client,
    });
  });

  afterEach(async () => {
    await graphqlCtx.after();
  });

  return ctx;
}

function graphqlTestContext() {
  let serverInstance: ServerInfo | null = null;

  return {
    async before() {
      const port = await getPort({ port: makeRange(4000, 6000) });

      serverInstance = await testServer.listen({ port });

      console.log;

      return new GraphQLClient(`http://localhost:${port}`);
    },
    async after() {
      serverInstance?.server.close();
    },
  };
}
