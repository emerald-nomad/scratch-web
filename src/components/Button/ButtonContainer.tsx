import { ReactElement } from "react";
import { ButtonProps } from "./Button";
import styles from "./Button.module.scss";

interface ButtonContainerProps {
  children: ReactElement<ButtonProps>[];
}

const ButtonContainer: React.FC<ButtonContainerProps> = ({ children }) => {
  return <div className={styles["btn-container"]}>{children}</div>;
};

export default ButtonContainer;
