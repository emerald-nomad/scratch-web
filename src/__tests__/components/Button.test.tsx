/**
 * @jest-environment jsdom
 */

import { render, screen } from "lib";
import Button, { ButtonProps } from "components/Button/Button";

describe("Button Component", () => {
  it("should have a 'btn--solid' class", () => {
    render(<Button text="Btn" style="solid" />);

    const btn = screen.getByRole("button");

    expect(btn.classList.contains("btn--solid"));
    expect(btn).toMatchSnapshot();
  });

  it("should have a 'btn--outline' class", () => {
    render(<Button text="Btn" style="outline" />);

    const btn = screen.getByRole("button");

    expect(btn.classList.contains("btn--outline"));
    expect(btn).toMatchSnapshot();
  });

  it("should have a 'btn--text' class", () => {
    render(<Button text="Btn" style="text" />);

    const btn = screen.getByRole("button");

    expect(btn.classList.contains("btn--text"));
    expect(btn).toMatchSnapshot();
  });

  it("should have the text 'Btn'", () => {
    render(<Button text="Btn" />);

    const btn = screen.getByRole("button");
    expect(screen.getByText("Btn"));
    expect(btn).toMatchSnapshot();
  });

  it("should show the loading indicator instead of text while loading", () => {
    render(<Button text="Btn" loading={true} />);

    const btn = screen.getByRole("button");

    expect(screen.getByLabelText("Loading Indicator"));
    expect(btn).toMatchSnapshot();
  });
});
