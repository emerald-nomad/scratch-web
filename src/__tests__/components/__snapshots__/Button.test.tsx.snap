// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component should have a 'btn--outline' class 1`] = `
<button
  class="btn btn--outline"
  type="button"
>
  Btn
</button>
`;

exports[`Button Component should have a 'btn--solid' class 1`] = `
<button
  class="btn btn--solid"
  type="button"
>
  Btn
</button>
`;

exports[`Button Component should have a 'btn--text' class 1`] = `
<button
  class="btn btn--text"
  type="button"
>
  Btn
</button>
`;

exports[`Button Component should have the text 'Btn' 1`] = `
<button
  class="btn btn--solid"
  type="button"
>
  Btn
</button>
`;

exports[`Button Component should show the loading indicator instead of text while loading 1`] = `
<button
  class="btn btn--solid"
  type="button"
>
  <div
    aria-label="Loading Indicator"
    class="lds-ring lds-ring--white"
  >
    <div />
    <div />
    <div />
    <div />
  </div>
</button>
`;
